File: 1691-minimum-number-of-days-to-disconnect-island/1691-minimum-number-of-days-to-disconnect-island.py
class Solution:
    def minDays(self, A: List[List[int]]) -> int:
        R, C = len(A), len(A[0])

        def neighbors(x, y):
            for r, c in ((x+1, y), (x,y+1), (x-1, y), (x,y-1)):
                if 0 <= r < R and 0 <= c < C and A[r][c] == 1:
                    yield r, c

        def islands():
            isl = 0
            seen = set()
            
            def dfs(r, c):
                for nr, nc in neighbors(r, c):
                    if (nr, nc) not in seen:
                        seen.add((nr, nc))
                        dfs(nr, nc)
                
            for r, c in product(range(R), range(C)):
                if (r, c) not in seen and A[r][c] == 1:
                    isl += 1
                    seen.add((r, c))
                    dfs(r, c)

            return isl != 1
        
        ones = sum(sum(row) for row in A)

        def find():
            ans = (-1, -1, -1)
            for r, c in product(range(R), range(C)):
                if A[r][c] == 0:
                    continue
                valid = len(list(neighbors(r, c)))
                ans = max(ans, (valid, r, c))
            
            _, r, c = ans
            A[r][c] = 0

        if islands():
            return 0

        for r, c in product(range(R), range(C)):
            if A[r][c] == 1:
                A[r][c] = 0
                if islands():
                    return 1
                A[r][c] = 1
        
        return 2

File: 1691-minimum-number-of-days-to-disconnect-island/1691-minimum-number-of-days-to-disconnect-island.java
class Solution {
    int[][] grid;
    
    public int minDays(int[][] grid) {
        int n = grid.length, m = grid[0].length;
        this.grid = grid;

        if (!isIsland()) {
            return 0;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    grid[i][j] = 0; 
                    if (!isIsland()) {
                        return 1;
                    }
                    grid[i][j] = 1;
                }
            }
        }

        return 2;
    }

    boolean isIsland() {
        int n = grid.length, m = grid[0].length;

        boolean[][] vis = new boolean[n][m];
        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!vis[i][j] && grid[i][j] == 1) {
                    count++;
                    if (count > 1) {
                        return false;
                    }
                    dfs(grid, vis, i, j);
                }
            }
        }

        return count == 1;
    }

    void dfs(int[][] mat, boolean[][] seen, int i, int j) {
        if (i < 0 || i >= mat.length || j < 0 || j >= mat[0].length || seen[i][j] || mat[i][j] == 0) {
            return;
        }
        seen[i][j] = true;

        dfs(mat, seen, i + 1, j);  // down
        dfs(mat, seen, i - 1, j);  // up
        dfs(mat, seen, i, j + 1);  // right
        dfs(mat, seen, i, j - 1);  // left
    }
}
